module/zfs/zfs_vnops.c

```diff
@@ -1105,6 +1106,16 @@ zfs_clone_range(znode_t *inzp, uint64_t *inoffp, znode_t *outzp,
 		return (error);
 	}
 
+	/*
+	 * Cloning across encrypted datasets is possible only if they
+	 * share the same master key.
+	 */
+	if (inos != outos && inos->os_encrypted &&
+	    !dmu_objset_crypto_key_equal(inos, outos)) {
+		zfs_exit_two(inzfsvfs, outzfsvfs, FTAG);
+		return (SET_ERROR(EXDEV));
+	}
```

Comment:
nit: I'd suggest moving this before the `zfs_verify_zp()` check.  This colocates it directly after the unencrypted dataset check and helps make it clear neither `inzp` or `outzp` are used.

Reply:
good point changed it.